libvmi/os/linux/elfparse.c

```diff
@@ -0,0 +1,422 @@
+             * so we cross-check our result with the string table.
+             */
+            __ctx.addr = str_tab + sym.st_name;
+            char *n = vmi_read_str(vmi, &__ctx);
+            if (!n || strcmp(name, n) != 0) {
+                if (n)
+                    free(n);
+                continue;
+            }
+
+            free(n);
+            *rva = sym.st_value;
+            return VMI_SUCCESS;
+        }
+
+        if (chain & 0b1)
+            break;
+    }
+
+    return VMI_FAILURE;
+}
+
+static uint32_t gnu_hash_max_symbols(
+    vmi_instance_t vmi,
+    const access_context_t *ctx,
+    addr_t hash_tab,
+    uint32_t *max_symbols)
+{
+    access_context_t _ctx = *ctx;
+    uint32_t table[4];
+
+    _ctx.addr = hash_tab;
+    if (VMI_FAILURE == vmi_read(vmi, &_ctx, sizeof(table), table, NULL))
+        return VMI_FAILURE;
+
+    uint32_t buckets[table[0]];
+    _ctx.addr += sizeof(table) + sizeof(uint64_t) * table[2];
+    if (VMI_FAILURE == vmi_read(vmi, &_ctx, sizeof(buckets), buckets, NULL))
+        return VMI_FAILURE;
+
+    uint32_t last = 0;
+    for (uint32_t i = 0; i < table[0]; i++)
+        if (buckets[i] > last)
+            last = buckets[i];
+
+    for (;; last++) {
+        uint32_t chain;
+        _ctx.addr = hash_tab + sizeof(table) + sizeof(uint64_t) * table[2]
+                    + sizeof(buckets) + sizeof(uint32_t) * (last - table[1]);
+        if (VMI_FAILURE == vmi_read_32(vmi, &_ctx, &chain))
+            return VMI_FAILURE;
+
+        if (chain & 0b1)
+            break;
+    }
+
+    *max_symbols = last;
+    return VMI_SUCCESS;
+}
+
+/* returns the rva value for a linux ELF dynamic symbol */
+status_t
+linux_export_to_rva(
+    vmi_instance_t vmi,
+    const access_context_t *ctx,
+    const char *symbol,
+    addr_t *rva)
+{
+    addr_t hash = 0, gnu_hash = 0, str = 0, sym = 0;
+    uint8_t image[1024] = { 0 };
+
+    if (VMI_FAILURE == elfparse_get_image(vmi, ctx, sizeof(image), image))
+        return VMI_FAILURE;
+
+    elfparse_assign_tables(vmi, ctx, image, &hash, &gnu_hash, &str, &sym);
+
+    if (!str || !sym || (!hash && !gnu_hash)) {
+        dbprint(VMI_DEBUG_ELFPARSE, "--ELFParse: failed to locate dynamic tables\n");
+        return VMI_FAILURE;
+    }
+
+    /* approach 1: attempt to resolve the symbol through DT_GNU_HASH. */
+    if (gnu_hash)
+        return gnu_hash_lookup(vmi, ctx, str, sym, gnu_hash, symbol, rva);
+
+    /*
+     * note that on modern systems, the fallback will fail.
+     * most executables and shared libraries will still contain DT_HASH,
+     * however, since the dynamic linker only uses DT_GNU_HASH, DT_HASH is never accessed.
+     * hence, the corresponding page is not present in main memory.
+     * thus, the fallback is only intended for legacy systems (2006 and earlier).
+     */
+    dbprint(VMI_DEBUG_ELFPARSE, "--ELFParse: falling back to DT_HASH resolving\n");
+
+    /* approach 2: attempt to resolve the symbol through DT_HASH. */
+    if (hash)
```

Comment:
Superfluous if-check, you can just `return elf_hash_lookup` at this point and get rid of the unreachable VMI_FAILURE further down.

Reply:
Good catch!